### PR TITLE
Improve Expr[T].show and enable DiagrammedExprMacro

### PR DESCRIPTION
**Pull request by liufengyun**
Some improvements due to latest Dotty change.

BTW, we have added scalatest 3.1.x to Dotty community build https://github.com/lampepfl/dotty/pull/6444.

### CODE REVIEW

**Comment by cheeseng** [comment]
Hi @liufengyun I submitted a PR to your PR here: 

https://github.com/liufengyun/scalatest/pull/3

with some comments.

Thanks!

**Comment by liufengyun** [comment]
@cheeseng With the last commit, diagrammed assertion works.

I just checked a small part of the tests, most of them work nicely.

File: scalatest.dotty/src/main/scala/org/scalatest/DiagrammedExprMacro.scala [hunk 1]
```diff
@@ -38,112 +38,128 @@ object DiagrammedExprMacro {
...
-              if (l.value) DiagrammedExpr.applyExpr(l, r :: Nil, r.value, ${ getAnchor(expr) })
+              if (l.value) DiagrammedExpr.applyExpr(l, r :: Nil, r.value, $anchor)
               else l
             }
           case _ =>
-            val left = parse(lhs.seal.cast[Any])
-            val right = parse(rhs.seal.cast[Any])
+            type S
+            implicit val tpS: quoted.Type[S] = lhs.tpe.seal.asInstanceOf[quoted.Type[S]]
+            val left = parse[S](lhs.seal.asInstanceOf[Expr[S]])
```
Comment (liufengyun) [inline scalatest.dotty/src/main/scala/org/scalatest/DiagrammedExprMacro.scala]: @nicolasstucki I have to use a few hacks like the above. I'm not sure if there is better ways to achieve the goal.

Comment (nicolasstucki) [reply scalatest.dotty/src/main/scala/org/scalatest/DiagrammedExprMacro.scala]: That was the goal of the old `Sealed` abstraction that we never merged. I will revisit it.

**Comment by liufengyun** [comment]
@cheeseng 70% of the tests are passing now. Is it possible to get this in and improve as we go?

**Comment by cheeseng** [comment]
@liufengyun Sorry for the late response, it is a long weekend over here due to public holidays, yes I am fine by merging this, unfortunately I think @bvenners is teaching this week though.

**Comment by liufengyun** [comment]
The latest status: 96% test passes, only 20 tests disabled, of which 6 are actual bugs (related to Dotty).

It depends on position fixes https://github.com/lampepfl/dotty/pull/6584

**Comment by bvenners** [comment]
@liufengyun @cheeseng I got four failing tests:

```
[info] DiagrammedAssertionsSpec:
[info] DiagrammedAssertions The assert(boolean) method 
[info] - should throw TestFailedException with correct message and stack depth when is used to check 0 == a *** FAILED *** (1 millisecond)
[info]   Some("
[info]   
[info]   assert(0 == a)
[info]          | |  |
[info]          0 |  3
[info]            false
[info]   ") was not equal to Some("
[info]   
[info]   assert(0 == a)
[info]          |    |  |
[info]          0 |  3
[info]               false
[info]   ") (DiagrammedAssertionsSpec.scala:270)
[info]   Analysis:
[info]   Some(value: "...rt(0 == a)
[info]        | [|  |
[info]        0 |  3
[info] ]         false
[info] " -> "...rt(0 == a)
[info]        | [   |  |
[info]        0 |  3
[info]    ]         false
[info] ")
[info] DiagrammedAssertionsSpec:
[info] DiagrammedAssertions The assert(boolean, clue) method 
[info] - should throw TestFailedException with correct message and stack depth when is used to check 0 == a *** FAILED *** (1 millisecond)
[info]   Some("this is a clue
[info]   
[info]   assert(0 == a, "this is a clue")
[info]          | |  |
[info]          0 |  3
[info]            false
[info]   ") was not equal to Some("this is a clue
[info]   
[info]   assert(0 == a, "this is a clue")
[info]          |    |  |
[info]          0 |  3
[info]               false
[info]   ") (DiagrammedAssertionsSpec.scala:2735)
[info]   Analysis:
[info]   Some(value: "...s a clue")
[info]        | [|  |
[info]        0 |  3
[info] ]         false
[info] " -> "...s a clue")
[info]        | [   |  |
[info]        0 |  3
[info]    ]         false
[info] ")
[info] DiagrammedAssertionsSpec:
[info] DiagrammedAssertions The assume(boolean) method 
[info] - should throw TestCanceledException with correct message and stack depth when is used to check 0 == a *** FAILED *** (1 millisecond)
[info]   Some("
[info]   
[info]   assume(0 == a)
[info]          | |  |
[info]          0 |  3
[info]            false
[info]   ") was not equal to Some("
[info]   
[info]   assume(0 == a)
[info]          |    |  |
[info]          0 |  3
[info]               false
[info]   ") (DiagrammedAssertionsSpec.scala:5205)
[info]   Analysis:
[info]   Some(value: "...me(0 == a)
[info]        | [|  |
[info]        0 |  3
[info] ]         false
[info] " -> "...me(0 == a)
[info]        | [   |  |
[info]        0 |  3
[info]    ]         false
[info] ")
[info] DiagrammedAssertionsSpec:
[info] DiagrammedAssertions The assume(boolean, clue) method 
[info] - should throw TestCanceledException with correct message and stack depth when is used to check 0 == a *** FAILED *** (1 millisecond)
[info]   Some("this is a clue
[info]   
[info]   assume(0 == a, "this is a clue")
[info]          | |  |
[info]          0 |  3
[info]            false
[info]   ") was not equal to Some("this is a clue
[info]   
[info]   assume(0 == a, "this is a clue")
[info]          |    |  |
[info]          0 |  3
[info]               false
[info]   ") (DiagrammedAssertionsSpec.scala:7675)
[info]   Analysis:
[info]   Some(value: "...s a clue")
[info]        | [|  |
[info]        0 |  3
[info] ]         false
[info] " -> "...s a clue")
[info]        | [   |  |
[info]        0 |  3
[info]    ]         false
[info] ")
```

**Comment by cheeseng** [comment]
@bvenners  You encountered those failing tests under 2.12 right?  I think @liufengyun probably is working on Dotty side only, I'll take a look now and see if I can get them green on both old and new side.

Cheers.

**Comment by cheeseng** [comment]
@liufengyun When I try to build your latest branch I encountered the following compile error when trying to build scalatest: 

```
[error] -- [E006] Unbound Identifier Error: /home/cheeseng/git/scalatest-3.1/scalatest.dotty/src/main/scala/org/scalatest/DiagrammedExprMacro.scala:136:14 
[error] 136 |              lets(diagrams) { rs =>
[error]     |              ^^^^
[error]     |              Not found: lets
[error] -- [E006] Unbound Identifier Error: /home/cheeseng/git/scalatest-3.1/scalatest.dotty/src/main/scala/org/scalatest/DiagrammedExprMacro.scala:156:10 
[error] 156 |          lets(diagrams) { rs =>
[error]     |          ^^^^
[error]     |          Not found: lets
[error] -- [E006] Unbound Identifier Error: /home/cheeseng/git/scalatest-3.1/scalatest.dotty/src/main/scala/org/scalatest/DiagrammedExprMacro.scala:195:10 
[error] 195 |          lets(diagrams) { rs =>
[error]     |          ^^^^
[error]     |          Not found: lets
```

Not sure if they are due to the new nightly build?  I'll go for the failing tests under 2.12 first.

**Comment by cheeseng** [comment]
@liufengyun @bvenners I submitted a fix PR for the failing tests here: 

https://github.com/liufengyun/scalatest/pull/5

If I can get the dotty to compiles again, I can try to enhance those tests (currently skipped because Dotty can do better job) to detect Scala/Dotty build and assert differently.

Cheers.

**Comment by liufengyun** [comment]
@bvenners @cheeseng I've switched to the nightly, it should fix the compilation problem and tests.

File: scalatest-test/src/test/scala/org/scalatest/DiagrammedAssertionsSpec.scala [hunk 2]
```diff
@@ -272,9 +272,9 @@ class DiagrammedAssertionsSpec extends FunSpec with Matchers with DiagrammedAsse
             """
               |
               |assert(0 == a)
-              |       |    |  |
+              |       | |  |
               |       0 |  3
-              |            false
+              |         false
               |""".stripMargin
           )
```
Comment (liufengyun) [inline scalatest-test/src/test/scala/org/scalatest/DiagrammedAssertionsSpec.scala]: @cheeseng Does it mean Scala2 compiler have somewhere improved the positions? Now Dotty agrees with Scala2 on these tests.

**Pull request merged**